compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -197,6 +235,137 @@ private void performCodeGen(boolean shrink) {
         codeGenCompleted = true;
     }
 
+    private void registerUnusedBIRNodes() {
+        // Reversed the for loop because used BIRNode analysis should start from the root module.
+        // Root module is usually found last in the topologicallySortedModuleList.
+        for (int i = pkgResolution.topologicallySortedModuleList().size() - 1; i >= 0; i--) {
+            ModuleContext moduleContext = pkgResolution.topologicallySortedModuleList().get(i);
+
+            // Default module is analyzed first to find its immediate dependencies.
+            // Its immediate dependent modules are marked as "used" and they are optimized after that.
+            // This process happens till all "used" modules are exhausted.
+            if (shouldOptimize(moduleContext) && (isRootModule(moduleContext) || moduleContext.isUsed())) {
+                this.usedBIRNodeAnalyzer.analyze(moduleContext.bLangPackage());
+                updateNativeDependencyMap(moduleContext);
+            }
+        }
+
+        CodeGenOptimizationReportEmitter.emitBirOptimizationDuration();
+
+        if (this.packageContext.project().buildOptions().verbose()) {
+            CodeGenOptimizationReportEmitter.emitCodegenOptimizationReport(
+                    this.usedBIRNodeAnalyzer.pkgWiseInvocationData, getOptimizationReportPath(),
+                    packageContext.project().kind());
+        }
+    }
+
+    private Path getOptimizationReportPath() {
+        return this.packageContext.project().sourceRoot;
+    }
+
+    protected Set<PackageID> getOptimizedPackageIDs() {
+        return this.usedBIRNodeAnalyzer.pkgWiseInvocationData.keySet();
+    }
+
+    private void markTestDependenciesForDuplicateBIRGen() {
+        for (int i = pkgResolution.topologicallySortedModuleList().size() - 1; i >= 0; i--) {
```

Comment:
Extract this to a variable `pkgResolution.topologicallySortedModuleList()`.

Reply:
Fixed in aa8254a61e2f3d88c1fdd944ffe5a5b257ede583.